ibm/resource_ibm_pi_image.go

```diff
@@ -63,145 +122,193 @@ func resourceIBMPIImage() *schema.Resource {
 	}
 }
 
-func resourceIBMPIImageCreate(d *schema.ResourceData, meta interface{}) error {
+func resourceIBMPIImageCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	sess, err := meta.(ClientSession).IBMPISession()
 	if err != nil {
 		log.Printf("Failed to get the session")
-		return err
+		return diag.FromErr(err)
 	}
 
-	powerinstanceid := d.Get(helpers.PICloudInstanceId).(string)
-	name := d.Get(helpers.PIImageName).(string)
-	imageid := d.Get(helpers.PIInstanceImageName).(string)
+	cloudInstanceID := d.Get(helpers.PICloudInstanceId).(string)
+	imageName := d.Get(helpers.PIImageName).(string)
 
-	client := st.NewIBMPIImageClient(sess, powerinstanceid)
+	client := st.NewIBMPIImageClient(sess, cloudInstanceID)
+	// image copy
+	if v, ok := d.GetOk(helpers.PIImageId); ok {
+		imageid := v.(string)
+		imageResponse, err := client.Create(imageName, imageid, cloudInstanceID)
+		if err != nil {
+			return diag.FromErr(err)
+		}
 
-	imageResponse, err := client.Create(name, imageid, powerinstanceid)
-	if err != nil {
-		return err
+		IBMPIImageID := imageResponse.ImageID
+		d.SetId(fmt.Sprintf("%s/%s", cloudInstanceID, *IBMPIImageID))
+
+		_, err = isWaitForIBMPIImageAvailable(ctx, client, *IBMPIImageID, d.Timeout(schema.TimeoutCreate), cloudInstanceID)
+		if err != nil {
+			log.Printf("[DEBUG]  err %s", err)
+			return diag.FromErr(err)
+		}
 	}
 
-	IBMPIImageID := imageResponse.ImageID
-	d.SetId(fmt.Sprintf("%s/%s", powerinstanceid, *IBMPIImageID))
+	// COS image import
+	if v, ok := d.GetOk(helpers.PIImageBucketName); ok {
+		bucketName := v.(string)
+		bucketImageFileName := d.Get(helpers.PIImageBucketFileName).(string)
+		bucketRegion := d.Get(helpers.PIImageBucketRegion).(string)
+		bucketAccess := d.Get(helpers.PIImageBucketAccess).(string)
+		storageType := d.Get(helpers.PIImageStorageType).(string)
+
+		body := &models.CreateCosImageImportJob{
+			ImageName:     &imageName,
+			BucketName:    &bucketName,
+			BucketAccess:  &bucketAccess,
+			ImageFilename: &bucketImageFileName,
+			Region:        &bucketRegion,
+			StorageType:   storageType,
+		}
 
-	_, err = isWaitForIBMPIImageAvailable(client, *IBMPIImageID, d.Timeout(schema.TimeoutCreate), powerinstanceid)
-	if err != nil {
-		log.Printf("[DEBUG]  err %s", err)
-		return err
+		if v, ok := d.GetOk(helpers.PIImageAccessKey); ok {
+			body.AccessKey = v.(string)
+		}
+		if v, ok := d.GetOk(helpers.PIImageSecretKey); ok {
+			body.SecretKey = v.(string)
+		}
+
+		imageResponse, err := client.CreateCosImageWithContext(ctx, body, cloudInstanceID)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		jobClient := st.NewIBMPIJobClient(sess, cloudInstanceID)
+		_, err = waitForIBMPIJobCompleted(ctx, jobClient, *imageResponse.ID, cloudInstanceID, d.Timeout(schema.TimeoutCreate))
+		if err != nil {
+			return diag.Errorf(errors.CreateImageOperationFailed, cloudInstanceID, err)
+		}
+
```

Comment:
Shoule we add the wait logic for the image to be uploaded from COS if its huge image does it take time? and wait for the status to be active or soemthing?

Reply:
Wait logic is applied for the job status. The API flow is asynchronous and job completion ensures the image is available in the target state.